pkg/snapshot/copier/copier.go

```diff
@@ -0,0 +1,128 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package copier
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/gardener/etcd-backup-restore/pkg/miscellaneous"
+	"github.com/gardener/etcd-backup-restore/pkg/snapstore"
+	brtypes "github.com/gardener/etcd-backup-restore/pkg/types"
+
+	"github.com/sirupsen/logrus"
+)
+
+// GetSourceAndDestinationStores returns the source and destination stores for the given source and destination and store configs.
+func GetSourceAndDestinationStores(sourceSnapStoreConfig *brtypes.SnapstoreConfig, destSnapStoreConfig *brtypes.SnapstoreConfig) (brtypes.SnapStore, brtypes.SnapStore, error) {
+	sourceSnapStore, err := snapstore.GetSnapstore(sourceSnapStoreConfig)
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not get source snapstore: %v", err)
+	}
+
+	destSnapStore, err := snapstore.GetSnapstore(destSnapStoreConfig)
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not get destination snapstore: %v", err)
+	}
+
+	return sourceSnapStore, destSnapStore, nil
+}
+
+// Copier can be used to copy backups from a source to a destination store.
+type Copier struct {
+	logger          *logrus.Entry
+	sourceSnapStore brtypes.SnapStore
+	destSnapStore   brtypes.SnapStore
+	maxBackups      int
+	maxBackupAge    int
+}
+
+// NewCopier creates a new copier.
+func NewCopier(sourceSnapStore brtypes.SnapStore, destSnapStore brtypes.SnapStore, logger *logrus.Entry, maxBackups int, maxBackupAge int) *Copier {
+	return &Copier{
+		logger:          logger.WithField("actor", "copier"),
+		sourceSnapStore: sourceSnapStore,
+		destSnapStore:   destSnapStore,
+		maxBackups:      maxBackups,
+		maxBackupAge:    maxBackupAge,
+	}
+}
+
+// Run executes the copy command.
+func (c *Copier) Run() error {
+	return c.CopyBackups()
+}
+
+// CopyBackups copies all backups from the source store to the destination store.
+func (c *Copier) CopyBackups() error {
+	// Get source backups
+	c.logger.Info("Getting source backups...")
+	sourceSnapshot, err := c.getSnapshots()
+	if err != nil {
+		return fmt.Errorf("could not get source backups: %v", err)
+	}
+
+	// If there are no source backups, do nothing
+	if len(sourceSnapshot) == 0 {
+		c.logger.Info("No source backups found")
+		return nil
+	}
+
+	// Get destination snapshots and build a map keyed by name
+	c.logger.Info("Getting destination snapshots...")
+	destSnapshots, err := c.destSnapStore.List()
+	if err != nil {
+		return fmt.Errorf("could not get destination snapshots: %v", err)
+	}
+	destSnapshotsMap := make(map[string]*brtypes.Snapshot)
+	for _, snapshot := range destSnapshots {
+		destSnapshotsMap[snapshot.SnapName] = snapshot
+	}
+
+	for _, snapshot := range sourceSnapshot {
+		// Copy all the snapshots (if not already copied)
+		if _, ok := destSnapshotsMap[snapshot.SnapName]; !ok {
+			c.logger.Infof("Copying %s snapshot %s...", snapshot.Kind, snapshot.SnapName)
+			if err := c.copySnapshot(snapshot); err != nil {
```

Comment:
We should think about parallelization in a future PR.

Reply:
Do you think this could increase the performance? I think the bottleneck is likely to be the read / write throughput of the backup container, and if this is the case then copying multiple snapshots in parallel is unlikely to be faster. But of course this is just an unverified assumption that should be checked.

Reply:
I'm not sure either but maybe worth a try :)